src/sync.cpp

```diff
@@ -2537,8 +2737,931 @@ void Syncs::resumeResumableSyncsOnStartup()
     }
 }
 
+JSONSyncConfig::JSONSyncConfig()
+  : drivePath()
+  , sourcePath()
+  , targetPath()
+  , fingerprint(0)
+  , targetHandle(UNDEF)
+  , lastError(NO_SYNC_ERROR)
+  , lastWarning(NO_SYNC_WARNING)
+  , type()
+  , backupId(UNDEF)
+  , enabled(false)
+{
+}
+
+bool JSONSyncConfig::external() const
+{
+    return !drivePath.empty();
+}
 
+bool JSONSyncConfig::valid() const
+{
+    return !(sourcePath.empty()
+             || targetPath.empty()
+             || targetHandle == UNDEF);
+}
 
+bool JSONSyncConfig::operator==(const JSONSyncConfig& rhs) const
+{
+    return drivePath == rhs.drivePath
+           && sourcePath == rhs.sourcePath
+           && targetPath == rhs.targetPath
+           && fingerprint == rhs.fingerprint
+           && targetHandle == rhs.targetHandle
+           && lastError == rhs.lastError
+           && lastWarning == rhs.lastWarning
+           && type == rhs.type
+           && backupId == rhs.backupId
+           && enabled == rhs.enabled;
+}
+
+bool JSONSyncConfig::operator!=(const JSONSyncConfig& rhs) const
+{
+    return !(*this == rhs);
+}
+
+JSONSyncConfig translate(const MegaClient& client, const SyncConfig &config)
+{
+    JSONSyncConfig result;
+
+    result.enabled = config.getEnabled();
+    result.fingerprint = config.getLocalFingerprint();
+    result.lastError = config.getError();
+    result.lastWarning = config.getWarning();
+    result.backupId = config.getBackupId();
+    result.targetHandle = config.getRemoteNode();
+    result.targetPath = config.getRemotePath();
+    result.type = config.getType();
+
+    const auto& drivePath = config.drivePath();
+    const auto& name = config.getName();
+    const auto sourcePath = config.getLocalPath().substr(drivePath.size());
+
+    const auto& fsAccess = *client.fsaccess;
+
+    result.drivePath = LocalPath::fromPath(drivePath, fsAccess);
+    result.sourcePath = LocalPath::fromPath(sourcePath, fsAccess);
+
+    // Ensure paths are normalized.
+    if (config.isExternal())
+    {
+        result.drivePath = NormalizeAbsolute(result.drivePath);
+        result.sourcePath = NormalizeRelative(result.sourcePath);
+    }
+    else
+    {
+        result.sourcePath = NormalizeAbsolute(result.sourcePath);
+    }
+
+    if (name == sourcePath)
+    {
+        result.name = result.sourcePath.toPath(fsAccess);
+    }
+    else
+    {
+        result.name = name;
+    }
```

Comment:
Yes, this will go away when we merge SyncConfig and JSONSyncConfig too.  Actually I already ran into this difference between the two, I'm changing the field in SyncConfig to be a LocalPath (consolidating changes from the JSON version and removing that one), and that is having some other effects, such as having to move it from types.h to sync.h so that LocalPath is already declared and usable, and adapting the places it was used to perform conversions - or in many cases, to remove the conversion that was there, as LocalPath is the more natural type for this field.